feat: add ifs_same_cond

Gianfranco99 opened the pull request:
Summary 📌
This PR Resolves: https://github.com/keep-starknet-strange/cairo-lint/issues/87

What it does:
This PR introduces a new lint that detects consecutive if statements with the same condition. When two if or else if blocks check the same condition, this could be an indication of a copy & paste error, and this lint helps prevent such mistakes.
All tests pass
![image](https://github.com/user-attachments/assets/4f2c7500-bf88-442b-b52b-106a6babe6c9)

0xLucqs commented on crates/cairo-lint-core/src/lints/ifs/ifs_same_cond.rs:
```diff
@@ -0,0 +1,54 @@
+    matches!(expr, Expr::FunctionCall(_))
+}
+
+fn get_condition_text(db: &dyn SyntaxGroup, condition: &Condition) -> Option<String> {
+    match condition {
+        Condition::Expr(expr) => Some(expr.as_syntax_node().get_text(db).to_string()),
+        Condition::Let(condition_let) => Some(condition_let.expr(db).as_syntax_node().get_text(db).to_string()),
+    }
+}
```
why does it return an option ?

0xLucqs commented on crates/cairo-lint-core/src/lints/ifs/ifs_same_cond.rs:
```diff
@@ -0,0 +1,54 @@
+
+pub const DUPLICATE_IF_CONDITION: &str = "Consecutive `if` with the same condition found.";
+
+fn is_function_call(expr: &Expr) -> bool {
+    matches!(expr, Expr::FunctionCall(_))
+}
```
i don't think this is necessary you can directly use what's inside the function

Gianfranco99 replied:
Do you mean using it like this 
```
      let is_if_function_call = match &if_condition {
                Condition::Expr(expr) =>  matches!(&expr.expr(db), Expr::FunctionCall(_)),
                Condition::Let(condition_let) => matches!(&condition_let.expr(db),Expr::FunctionCall(_)),
            };

            let is_else_if_function_call = match &else_if_condition {
                Condition::Expr(expr) => matches!(&expr.expr(db), Expr::FunctionCall(_)),
                Condition::Let(condition_let) => matches!(&condition_let.expr(db),Expr::FunctionCall(_)),
            };

            if is_if_function_call || is_else_if_function_call {
                return;
            }
```

Gianfranco99 replied:
or do you like it better like this?
```
if matches!(&if_condition, Condition::Expr(expr) if matches!(expr.expr(db), Expr::FunctionCall(_)))
    || matches!(&if_condition, Condition::Let(condition_let) if matches!(condition_let.expr(db), Expr::FunctionCall(_)))
{
    return; // Salimos si la condición es una llamada a función
}
```

Gianfranco99 replied:
Now that I think about it, I think we can only validate with the if if it is a function or not, right? because if the if is not a function but the else_if is, they would no longer match and it would not apply to the case

0xLucqs commented on crates/cairo-lint-core/src/lints/ifs/ifs_same_cond.rs:
```diff
@@ -0,0 +1,54 @@
+    }
+}
+
+pub fn check_duplicate_if_condition(db: &dyn SyntaxGroup, if_expr: &ExprIf, diagnostics: &mut Vec<PluginDiagnostic>) {
```
you can have multiple `else if` the code doesn't seem to handle that case

0xLucqs commented on crates/cairo-lint-core/tests/test_files/ifs/ifs_same_cond:
```diff
@@ -0,0 +1,167 @@
+//! > Same condition with boolean
```
add a tests with:
- more than one else if
- one where 2 conditions are similar but not the last one
and try to think about more edge cases to avoid bugs

0xLucqs commented on crates/cairo-lint-core/tests/test_files/ifs/ifs_same_cond:
```diff
@@ -0,0 +1,370 @@
+//! > Combined conditions with different if
```
what if you use literals in the if condition ? `if 1 == 1` `if a == 3`

0xLucqs reviewed (changes requested):
add these tests otherwise looks good

0xLucqs commented:
any news ?

Gianfranco99 commented:
> any news ?
 
Done,  I'm going to do a push now. @0xLucqs

0xLucqs commented on crates/cairo-lint-core/src/lints/ifs/ifs_same_cond.rs:
```diff
@@ -0,0 +1,88 @@
+    }
+}
+
+fn are_conditions_equivalent(condition1: &str, condition2: &str) -> bool {
+    condition1.contains("==")
+        && condition2.contains("==")
+        && condition1.split("==").nth(0).unwrap().trim() == condition2.split("==").nth(1).unwrap().trim()
+        && condition1.split("==").nth(1).unwrap().trim() == condition2.split("==").nth(0).unwrap().trim()
+}
```
What are you trying to achieve here ?

Gianfranco99 replied:
What I do here is validate if the conditions of the if and the else if are the same even if they are changed in the order in the else if
example => `fn main(){
    let a = 3;

    if a == 3 {
        println!("a == 3");
    } else if 3 == a {
        println!("3 == a");
    }
}`

0xLucqs replied:
i mean you can just put the variables in a hashset and check its length ?

0xLucqs commented on crates/cairo-lint-core/src/lints/ifs/ifs_same_cond.rs:
```diff
@@ -0,0 +1,88 @@
+                let lhs_text = binary_expr.lhs(db).as_syntax_node().get_text(db).to_string();
+                let rhs_text = binary_expr.rhs(db).as_syntax_node().get_text(db).to_string();
+
+                match operator {
+                    BinaryOperator::Eq(_) => {
+                        return format!("{} == {}", lhs_text, rhs_text);
+                    }
+                    BinaryOperator::LT(_) => {
+                        return format!("({}, {})", rhs_text, lhs_text);
+                    }
+                    BinaryOperator::GT(_) => {
+                        return format!("({}, {})", lhs_text, rhs_text);
+                    }
+                    _ => {}
```
what is that ?

Gianfranco99 replied:
this is to match the conditions = , < , > 
example =>
fn main(){
    let a:u32 = 3;

    if a > 3 {
        println!("a == 3");
    } else if 3 < a {
        println!("3 == a");
    }
}

0xLucqs replied:
what do you do with the output, for example the function returns `(1, a)` how is that helpful ?

0xLucqs reviewed (changes requested):
still need the `1==1` test

Gianfranco99 commented on crates/cairo-lint-core/src/lints/ifs/ifs_same_cond.rs:
```diff
@@ -0,0 +1,88 @@
+                continue;
+            }
+
+            let if_condition_text = get_condition_text(db, &if_condition);
```
You're right, I can leave it out of the while.

Gianfranco99 commented:
> still need the `1==1` test

I had added that test but I deleted it since there was a conflict with a new test that validates the same condition and it generated conflicts in the file showing another error message, but if you want we can see how we solve this and also add that test

0xLucqs commented:
@all-contributors please add @Gianfranco99 for code and tests
